src/connection.rs

```diff
@@ -28,6 +28,22 @@ pub(crate) enum Inner {
     Server(ServerSession),
 }
 
+impl Conn {
+    fn as_ref(&self) -> &dyn Session {
```

Comment:
Maybe these should become `AsRef` and `AsMut` implementations?

Reply:
I tried this and got lifetime problems I couldn't figure out the source of:

```rust
impl AsRef<dyn Session> for Conn {
    fn as_ref(&self) -> &dyn Session {
        match &self.conn {
            Inner::Client(c) => c,
            Inner::Server(c) => c,
        }
    }
}

impl AsMut<dyn Session> for Conn {
    fn as_mut(&mut self) -> &mut dyn Session {
        match &mut self.conn {
            Inner::Client(c) => c,
            Inner::Server(c) => c,
        }
    }
}
```

```
error: `impl` item signature doesn't match `trait` item signature
   --> src/connection.rs:32:5
    |
32  |     fn as_ref(&self) -> &dyn Session {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found `fn(&Conn) -> &dyn Session`
    | 
   ::: /home/jsha/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/convert/mod.rs:158:5
    |
158 |     fn as_ref(&self) -> &T;
    |     ----------------------- expected `fn(&Conn) -> &(dyn Session + 'static)`
    |
    = note: expected `fn(&Conn) -> &(dyn Session + 'static)`
               found `fn(&Conn) -> &dyn Session`
    = help: the lifetime requirements from the `impl` do not correspond to the requirements in the `trait`
    = help: verify the lifetime relationships in the `trait` and `impl` between the `self` argument, the other inputs and its output

error: `impl` item signature doesn't match `trait` item signature
   --> src/connection.rs:41:5
    |
41  |     fn as_mut(&mut self) -> &mut dyn Session {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found `fn(&mut Conn) -> &mut dyn Session`
    | 
   ::: /home/jsha/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/convert/mod.rs:199:5
    |
199 |     fn as_mut(&mut self) -> &mut T;
    |     ------------------------------- expected `fn(&mut Conn) -> &mut (dyn Session + 'static)`
    |
    = note: expected `fn(&mut Conn) -> &mut (dyn Session + 'static)`
               found `fn(&mut Conn) -> &mut dyn Session`
    = help: the lifetime requirements from the `impl` do not correspond to the requirements in the `trait`
    = help: verify the lifetime relationships in the `trait` and `impl` between the `self` argument, the other inputs and its output
```

If you have any idea what's wrong here I'd be happy to switch over to `AsRef` / `AsMut`.

Reply:
Did you try adding the `+ 'static`?